internal/tar_ball_queue.go

```diff
@@ -0,0 +1,135 @@
+package internal
+
+import (
+	"github.com/pkg/errors"
+	"sync"
+	"sync/atomic"
+)
+
+// TarBallQueue is used to process multiple tarballs concurrently
+type TarBallQueue struct {
+	tarballQueue     chan TarBall
+	uploadQueue      chan TarBall
+	parallelTarballs int
+	maxUploadQueue   int
+	mutex            sync.Mutex
+	started          bool
+
+	TarSizeThreshold   int64
+	AllTarballsSize    *int64
+	TarBallMaker       TarBallMaker
+	LastCreatedTarball TarBall
+}
+
+func newTarBallQueue(tarSizeThreshold int64, tarBallMaker TarBallMaker) *TarBallQueue {
+	return &TarBallQueue{
+		TarSizeThreshold: tarSizeThreshold,
+		TarBallMaker:     tarBallMaker,
+		AllTarballsSize:  new(int64),
+	}
+}
+
+func (tarQueue *TarBallQueue) StartQueue() error {
+	if tarQueue.started {
+		panic("Trying to start already started Queue")
+	}
+	var err error
+	tarQueue.parallelTarballs, err = getMaxUploadDiskConcurrency()
+	if err != nil {
+		return err
+	}
+	tarQueue.maxUploadQueue, err = getMaxUploadQueue()
+	if err != nil {
+		return err
+	}
+
+	tarQueue.tarballQueue = make(chan TarBall, tarQueue.parallelTarballs)
+	tarQueue.uploadQueue = make(chan TarBall, tarQueue.parallelTarballs+tarQueue.maxUploadQueue)
+	for i := 0; i < tarQueue.parallelTarballs; i++ {
+		tarQueue.NewTarBall(true)
+		tarQueue.tarballQueue <- tarQueue.LastCreatedTarball
+	}
+
+	tarQueue.started = true
+	return nil
+}
+
+func (tarQueue *TarBallQueue) Deque() TarBall {
+	if !tarQueue.started {
+		panic("Trying to deque from not started Queue")
+	}
+	return <-tarQueue.tarballQueue
+}
+
+func (tarQueue *TarBallQueue) FinishQueue() error {
+	if !tarQueue.started {
+		panic("Trying to stop not started Queue")
+	}
+	tarQueue.started = false
```

Comment:
which methods could be invoked in a goroutine? I see you access this variable in several places and it looks like there is no synchronization to access it.

Reply:
I can add this information in methods documentation

Reply:
I've replaced default bool with atomic bool in 2d464f4